lib/seam/errors.rb

```diff
@@ -1,68 +1,73 @@
-      end
+    def initialize(error, status_code, request_id)
+      super(error[:message])
+      @code = error[:type]
+      @status_code = status_code
+      @request_id = request_id
+      @data = error[:data]
     end
+  end
 
-    class SeamHttpUnauthorizedError < SeamHttpApiError
-      def initialize(request_id)
-        super({type: "unauthorized", message: "Unauthorized"}, 401, request_id)
-      end
+  class HttpUnauthorizedError < HttpApiError
+    def initialize(request_id)
+      super({type: "unauthorized", message: "Unauthorized"}, 401, request_id)
     end
+  end
 
-    class SeamHttpInvalidInputError < SeamHttpApiError
-      attr_reader :validation_errors
+  class InvalidInputError < HttpApiError
+    attr_reader :validation_errors
 
-      def initialize(error, status_code, request_id)
-        super(error, status_code, request_id)
-        @code = "invalid_input"
-        @validation_errors = error["validation_errors"] || {}
-      end
+    def initialize(error, status_code, request_id)
+      super(error, status_code, request_id)
+      @code = "invalid_input"
+      @validation_errors = error["validation_errors"] || {}
+    end
 
-      def get_validation_error_messages(param_name)
-        @validation_errors.dig(param_name, "_errors") || []
-      end
+    def get_validation_error_messages(param_name)
+      @validation_errors.dig(param_name, "_errors") || []
     end
+  end
 
-    # Action attempt
-    class SeamActionAttemptError < StandardError
-      attr_reader :action_attempt
+  # Action attempt
+  class ActionAttemptError < StandardError
+    attr_reader :action_attempt
 
-      def initialize(message, action_attempt)
-        super(message)
-        @action_attempt = action_attempt
-      end
+    def initialize(message, action_attempt)
+      super(message)
+      @action_attempt = action_attempt
+    end
 
-      def name
-        self.class.name
-      end
+    def name
+      self.class.name
     end
+  end
 
-    class SeamActionAttemptFailedError < SeamActionAttemptError
-      attr_reader :code
+  class ActionAttemptFailedError < ActionAttemptError
+    attr_reader :code
 
-      def initialize(action_attempt)
-        super(action_attempt.error.message, action_attempt)
-        @code = action_attempt.error.type
-      end
+    def initialize(action_attempt)
+      super(action_attempt.error.message, action_attempt)
+      @code = action_attempt.error.type
     end
+  end
+
+  class ActionAttemptTimeoutError < ActionAttemptError
+    def initialize(action_attempt, timeout)
+      message = "Timed out waiting for action attempt after #{timeout}s"
+      super(message, action_attempt)
+    end
+  end
+end
 
-    class SeamActionAttemptTimeoutError < SeamActionAttemptError
-      def initialize(action_attempt, timeout)
-        message = "Timed out waiting for action attempt after #{timeout}s"
-        super(message, action_attempt)
-      end
+module Http
+  class SeamInvalidTokenError < StandardError
```

Comment:
Yea, it looks like if we wanted to do this we would need to make `Http` a module and then nest everything under there, like the main class, etc.I don't know if this is worth the complexity.

I found this: https://medium.com/@blazejkosmowski/ruby-opinions-using-private-constant-592553ec49d2

Seems like we can just mark most of the stuff inside the `Seam` module with `private_constant`. So the `HttpSeamInvalidTokenError` errors and `Logger`, etc can all be `private_constant`. This way, the module only exposes the public classes, errors, and methods we want to expose, e.g., `Seam::Http`, `Seam::HttpMultiWorkpsace`, `Seam::Webhook`, `Seam::ActionAttemptTimeoutError`, `Seam.from_api_key`, etc.

Reply:
I'm not sure I understand how `private_constant` will allow us to use `Seam::Http::SeamInvalidTokenError` given the conflict between the `Http` class and module. If we mark the class as private, will it not conflict with the module?

What if we rename the `Http` class to `HttpClient` so that it doesn't conflict with the `Http` module?

Reply:
I meant that since we can't do `Seam::Http::SeamInvalidTokenError` without significant internal refactor (introducing `Seam::Http::SingleWorkspace` or something), we can just put the errors in a `private_constant` internal module or as top level `private_constant` like `Seam::Errors::InvalidTokenError`.

Basically, we only need to ensure the public API is limited to what we want to expose. I'm not sure which of these is a "better" internal structure: `Seam::Http::SeamInvalidTokenError` or `Seam::Errors::SeamInvalidTokenError` or `Seam::SeamInvalidTokenError`. But it doesn't matter much at the moment, as we need to prioritize the public API.

> What if we rename the Http class to HttpClient so that it doesn't conflict with the Http module?

Hmm, so we have `Seam::Http::Client` / `Seam::Http::InvalidTokenError` etc and then `Seam::Client` and `Seam::MultiWorkspaceClient`? That would make sense to me.

Reply:
Actually, instead of `Seam::Http::Client` (which is private anyway) I expect we would just use a lib anyway, so `Seam::Http::Client` would go away in favor of whatever actual lib we use (similar to how in the JS SDK, `seam.client` points directly to Axios, NOT a wrapper class around Axios)

Reply:
We don't have `Seam::Http::Client`, we have just `Seam::Http`. As of https://github.com/seamapi/ruby-next/pull/89, `Client` was renamed to `Http` and `ClientMultiWorkspace` became `HttpMultiWorkspace`.

Reply:
If we mark `InvalidTokenError` private, we won't be able to use it outside the place where it's defined, for ex. in a test.

Reply:
Ok. So we should do `Seam:: Client` and `Seam:: ClientMultiWorkspace` and Seam::Http::`?

Reply:
If we rename back to `Client` and `ClientMultiWorkspace`, what `Seam::Http` will be then?

Reply:
Just a module containing all of the http "support" functions, classes, errors, etc. basically whatever we need to make `seam.client` work.